kube-controllers/pkg/controllers/loadbalancer/loadbalancer_controller.go

```diff
@@ -521,45 +497,45 @@ func (c *loadBalancerController) syncService(svcObj serviceObject) error {
 }
 
 // needsIPsAssigned determines if service IPFamilyPolicy is requirement is fulfilled by number of assigned IPs in IPAM storage
-func (c *loadBalancerController) needsIPsAssigned(svc *v1.Service, svcObj serviceObject) bool {
+func (c *loadBalancerController) needsIPsAssigned(svc *v1.Service, svcKey serviceKey) bool {
 	switch *svc.Spec.IPFamilyPolicy {
 	case v1.IPFamilyPolicySingleStack:
-		return len(c.allocationTracker.ipsByService[svcObj]) == 1
+		return len(c.allocationTracker.ipsByService[svcKey]) != 1
 	case v1.IPFamilyPolicyRequireDualStack, v1.IPFamilyPolicyPreferDualStack:
-		return len(c.allocationTracker.ipsByService[svcObj]) == 2
+		return len(c.allocationTracker.ipsByService[svcKey]) != 2
 	default:
 		return true
 	}
 }
 
 // needsStatusUpdate checks if the service needs status update.
 // service needs status update if the IPs in our IPAM tracker do not match what's in the status of the service
-func (c *loadBalancerController) needsStatusUpdate(svc *v1.Service, svcObj serviceObject) bool {
-	if len(svc.Status.LoadBalancer.Ingress) != len(c.allocationTracker.ipsByService[svcObj]) {
+func (c *loadBalancerController) needsStatusUpdate(svc *v1.Service, svcKey serviceKey) bool {
+	if len(svc.Status.LoadBalancer.Ingress) != len(c.allocationTracker.ipsByService[svcKey]) {
 		return true
 	}
 
+	needsStatusUpdate := false
 	for _, ingress := range svc.Status.LoadBalancer.Ingress {
-		if _, ok := c.allocationTracker.ipsByService[svcObj][ingress.IP]; !ok {
-			break
+		if _, ok := c.allocationTracker.ipsByService[svcKey][ingress.IP]; !ok {
+			needsStatusUpdate = true
 		}
-		return false
 	}
```

Comment:
We also should iterate the `ipsByService` and check that they are all present in the status, not just that all of the IPs in the status are allocated in IPAM.

Reply:
I think the check to see if length of `ipsByService` and the `Status.LoadBalancer.Ingress` is the same should prevent us from having to iterate over the ipsByService